File: Contents/InitialConsonantSearch/InitialConsonantSearch/Main/TEST/Cell/LineSettingCell.m
//
//  ChartSettingCell.m
//  ChartTypeTest
//
//  Created by Kwan Hyun Son on 2023/09/01.
//

#import "LineSettingCell.h"
#import "GeometricShapesView.h"

@interface LineSettingCell ()
@property (weak, nonatomic) IBOutlet UIButton *selectionBtn; // 단순한 라벨일 뿐이다. 하나 제외.
@property (weak, nonatomic) IBOutlet UIView *contentContainer;

@property (weak, nonatomic) IBOutlet MGUBrightnessBorderButton *colorBtn;
@property (weak, nonatomic) IBOutlet UIButton *boldButton;
@property (weak, nonatomic) IBOutlet MGUDropdownButton *dropButton;
@property (weak, nonatomic) IBOutlet UISwitch *toggleSwitch;
@property (weak, nonatomic) IBOutlet UIButton *textInputButton;

@property (weak, nonatomic) IBOutlet GeometricShapesView *geometricShapesView;
@end

@implementation LineSettingCell

- (void)awakeFromNib {
    [super awakeFromNib];
    if (self.selectionBtn != nil) {
        self.selectionBtn.titleLabel.adjustsFontSizeToFitWidth = YES;
        self.selectionBtn.titleLabel.minimumScaleFactor = 0.5;
        
        self.selectionBtn.tintAdjustmentMode = UIViewTintAdjustmentModeNormal;
        self.selectionBtn.userInteractionEnabled = NO; // Label 형태로 사용. color fill 버튼 제외
        self.selectionBtn.translatesAutoresizingMaskIntoConstraints = NO;
        NSLayoutConstraint *constraint = [self.selectionBtn.heightAnchor constraintEqualToConstant:44.0]; // 키움: 55.0
        constraint.priority = UILayoutPriorityDefaultHigh;
        constraint.active = YES;
        [self.selectionBtn setTitleColor:[UIColor blackColor] forState:UIControlStateNormal];
    }
    CommonInit(self);
}


#pragma mark - 생성 & 소멸

static void CommonInit(LineSettingCell *self) {
    self.contentContainer.backgroundColor = [UIColor clearColor];
    if ([self.reuseIdentifier isEqualToString:LineSettingCellIDColorFill] == YES) {
        [self setupColorFillMode];
    } else if ([self.reuseIdentifier isEqualToString:LineSettingCellIDColor] == YES) {
        [self setupColorMode];
    } else if ([self.reuseIdentifier isEqualToString:LineSettingCellIDColorDrop] == YES) {
        [self setupColorDropMode];
    } else if ([self.reuseIdentifier isEqualToString:LineSettingCellIDToggle] == YES) {
        [self setupToggleMode];
    } else if ([self.reuseIdentifier isEqualToString:LineSettingCellIDTextInput] == YES) {
        [self setupTextInputMode];
    } else if ([self.reuseIdentifier isEqualToString:LineSettingCellIDDrop] == YES) {
        [self setupDropMode];
    } else if ([self.reuseIdentifier isEqualToString:LineSettingCellIDFont] == YES) {
        [self setupFontMode];
    } else if ([self.reuseIdentifier isEqualToString:LineSettingCellIDGeometricShapes] == YES) {
        [self setupGeometricShapesMode];
    } else {
        
    }
}

- (void)setupColorFillMode { // 좌측 타이틀 체크 + 칼라버튼
    self.colorBtn.backgroundColor = [UIColor systemMintColor];
    self.colorBtn.layer.cornerRadius = 4.0;
    self.selectionBtn.userInteractionEnabled = YES; // Label 형태로 사용. color fill 버튼 제외
    if (@available(iOS 13, *)) {
        if (@available(iOS 15, *)) {
            UIButtonConfiguration *configuration = self.selectionBtn.configuration;
            if (configuration != nil) {
                configuration.baseBackgroundColor = [UIColor clearColor];
                self.selectionBtn.configuration = configuration;
            }
        }
        // UIImageSymbolConfiguration *config = [UIImageSymbolConfiguration configurationWithScale:UIImageSymbolScaleMedium];
        // selectedImage = [UIImage systemImageNamed:@"checkmark.circle.fill" withConfiguration:config];
        // normalImage = [UIImage systemImageNamed:@"checkmark.circle" withConfiguration:config];
    } else {
        self.selectionBtn.imageView.contentMode = UIViewContentModeCenter;
        UIColor *tintColor = [UIColor colorWithRed:15.0/255.0 green:40.0/255.0 blue:71.0/255.0 alpha:1.0];
        self.selectionBtn.tintColor = tintColor;
        self.selectionBtn.imageView.tintColor = tintColor;
        // NSBundle *designTimeBundle = [NSBundle bundleForClass:[self classForCoder]];
        // selectedImage = [UIImage imageNamed:@"checkmark.circle.fill"
        //                                     inBundle:designTimeBundle
        //                compatibleWithTraitCollection:nil];
        // selectedImage = [selectedImage imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate];
        // normalImage = [UIImage imageNamed:@"uncheckmark.circle"
        //                                   inBundle:designTimeBundle
        //              compatibleWithTraitCollection:nil];
        // normalImage = [normalImage imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate];
    }
    NSBundle *designTimeBundle = [NSBundle bundleForClass:[self classForCoder]];
    UIImage *selectedImage = [UIImage imageNamed:@"checkmark.circle.fill"
                                        inBundle:designTimeBundle
                   compatibleWithTraitCollection:nil];
    selectedImage = [selectedImage imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate];
    UIImage *normalImage = [UIImage imageNamed:@"uncheckmark.circle"
                                      inBundle:designTimeBundle
                 compatibleWithTraitCollection:nil];
    normalImage = [normalImage imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate];
    [self.selectionBtn setImage:selectedImage forState:UIControlStateSelected];
    [self.selectionBtn setImage:normalImage forState:UIControlStateNormal];
    [self.selectionBtn setTitleColor:[UIColor blackColor] forState:UIControlStateSelected];
    [self.selectionBtn setTitleColor:[UIColor lightGrayColor] forState:UIControlStateNormal];
}

- (void)setupColorMode {
    self.colorBtn.backgroundColor = [UIColor systemMintColor];
    self.colorBtn.layer.cornerRadius = 4.0;
}

- (void)setupColorDropMode {
    self.colorBtn.backgroundColor = [UIColor systemMintColor];
    self.colorBtn.layer.cornerRadius = 4.0;
    self.dropButton.backgroundColor = [UIColor clearColor];
    
    self.dropButton.cellClass = [MGULineWidthDropdownCell class];
    self.dropButton.dropdownData = @[@(0), @(1), @(2), @(3), @(4), @(5), @(6), @(7), @(8), @(9)];
    self.dropButton.selectedIndex = 0;
}

- (void)setupToggleMode {
}

- (void)setupTextInputMode {
    self.textInputButton.backgroundColor = [UIColor whiteColor];
    self.textInputButton.layer.borderColor = [UIColor lightGrayColor].CGColor;
    self.textInputButton.layer.borderWidth = 1.0 / [UIScreen mainScreen].scale;
    self.textInputButton.layer.cornerRadius = 5.0;
    [self.textInputButton setTitleColor:[UIColor systemBlueColor] forState:UIControlStateNormal];
    [self.textInputButton setTitleColor:[UIColor grayColor] forState:UIControlStateDisabled];
    self.textInputButton.titleLabel.adjustsFontSizeToFitWidth = YES;
    self.textInputButton.titleLabel.minimumScaleFactor = 0.5;
}

- (void)setupDropMode {
    self.dropButton.backgroundColor = [UIColor clearColor];
    self.dropButton.cellClass = [MGUTextDropdownCell class];
    self.dropButton.dropdownData = @[@"목업0", @"목업1", @"목업2"];
    self.dropButton.selectedIndex = 1;
    self.dropButton.placeHolderTextAlignment = NSTextAlignmentCenter;
    self.dropButton.textAlignment = NSTextAlignmentRight;
}

- (void)setupFontMode {
    self.boldButton.backgroundColor = [UIColor clearColor];
    self.boldButton.layer.cornerRadius = 4.0; // 4.0
    self.dropButton.backgroundColor = [UIColor clearColor];
    self.dropButton.cellClass = [MGUTextDropdownCell class];
    NSMutableArray <NSString *>*dropdownData = @[].mutableCopy;
    for (int i = 10; i <= 50; i++) {
        NSString *title = [NSString stringWithFormat:@"%d", i];
        [dropdownData addObject:title];
    }
    self.dropButton.dropdownData = dropdownData;
    self.dropButton.layer.cornerRadius = 4.0;
    self.dropButton.selectedIndex = 7;
    self.dropButton.placeHolderTextAlignment = NSTextAlignmentCenter;
    self.dropButton.textAlignment = NSTextAlignmentCenter;
}

- (void)setupGeometricShapesMode {
}

#pragma mark - Setter & Getter

- (void)setData:(DTOLineSetting *)model {
    _data = model;
    [self.selectionBtn setTitle:model.title forState:UIControlStateNormal];
    
    if ([self.reuseIdentifier isEqualToString:LineSettingCellIDColorFill] == YES) {
        self.colorBtn.backgroundColor = model.color;
        if (self.selectionBtn.isSelected != model.selected) {
            self.selectionBtn.selected = model.selected;
        }
    } else if ([self.reuseIdentifier isEqualToString:LineSettingCellIDColor] == YES) {
        self.colorBtn.backgroundColor = model.color;
    } else if ([self.reuseIdentifier isEqualToString:LineSettingCellIDColorDrop] == YES) {
        self.colorBtn.backgroundColor = model.color;
        self.dropButton.selectedIndex = model.dropBtnSelectedIndex;
    } else if ([self.reuseIdentifier isEqualToString:LineSettingCellIDToggle] == YES) {
        self.toggleSwitch.on = model.isToggleOn;
    } else if ([self.reuseIdentifier isEqualToString:LineSettingCellIDTextInput] == YES) {
        [self.textInputButton setTitle:model.textInputBtnTitle forState:UIControlStateNormal];
    } else if ([self.reuseIdentifier isEqualToString:LineSettingCellIDDrop] == YES) {
        if (model.dropBtnTitles != nil && model.dropBtnTitles.count > 0) {
            self.dropButton.dropdownData = model.dropBtnTitles;
        }
        self.dropButton.selectedIndex = model.dropBtnSelectedIndex;
    } else if ([self.reuseIdentifier isEqualToString:LineSettingCellIDFont] == YES) {
        if (self.boldButton.isSelected != model.isBold) {
            self.boldButton.selected = model.isBold;
        }
        if (self.dropButton.selectedIndex != model.dropBtnSelectedIndex) {
            self.dropButton.selectedIndex = model.dropBtnSelectedIndex;
        }
    } else if ([self.reuseIdentifier isEqualToString:LineSettingCellIDGeometricShapes] == YES) {
        self.geometricShapesView.geometricShapesType = self.data.geometricShapesType;
        self.geometricShapesView.borderColor = self.data.geometricShapesBorderColor;
        self.geometricShapesView.borderWidth = self.data.geometricShapesBorderWidth;
        self.geometricShapesView.backColor = self.data.geometricShapesBackColor;
    }
}

#pragma mark - Actions

@end

File: Contents/InitialConsonantSearch/IndicatorSettingsTest_Swift/Main/TEST/Cell/LineSettingCell.swift
//
//  ChartSettingCell.swift
//  ChartType_Swift
//
//  Created by Kwan Hyun Son on 2023/09/08.
//

import UIKit
import IosKit

final class LineSettingCell: UITableViewCell {
    
    // MARK: - Property
    
    weak var data: DTOLineSetting? {
        didSet {
            guard let model = data
            else {
                return
            }
            
            if let selectionBtn = selectionBtn {
                selectionBtn.setTitle(model.title, for: .normal)
            }
            
            if reuseIdentifier == LineSettingCellID.colorFill.rawValue {
                colorBtn.backgroundColor = model.color
                if let selectionBtn = selectionBtn, selectionBtn.isSelected != model.isSelected {
                    selectionBtn.isSelected = model.isSelected
                }
            } else if reuseIdentifier == LineSettingCellID.color.rawValue {
                colorBtn.backgroundColor = model.color
            } else if reuseIdentifier == LineSettingCellID.colorDrop.rawValue {
                colorBtn.backgroundColor = model.color
                dropButton.selectedIndex = model.dropBtnSelectedIndex
            } else if reuseIdentifier == LineSettingCellID.toggle.rawValue {
                toggleSwitch.isOn = model.isToggleOn
            } else if reuseIdentifier == LineSettingCellID.textInput.rawValue {
                textInputButton.setTitle(model.textInputBtnTitle, for: .normal)
            } else if reuseIdentifier == LineSettingCellID.drop.rawValue {
                if let dropBtnTitles = model.dropBtnTitles, dropBtnTitles.count > 0 {
                    dropButton.dropdownData = dropBtnTitles
                }
                dropButton.selectedIndex = model.dropBtnSelectedIndex
            } else if reuseIdentifier == LineSettingCellID.font.rawValue {
                if boldButton.isSelected != model.isBold {
                    boldButton.isSelected = model.isBold
                }
                if dropButton.selectedIndex != model.dropBtnSelectedIndex {
                    dropButton.selectedIndex = model.dropBtnSelectedIndex
                }
            } else if reuseIdentifier == LineSettingCellID.geometricShapes.rawValue {
                geometricShapesView.geometricShapesType = model.geometricShapesType
                geometricShapesView.borderColor = model.geometricShapesBorderColor ?? .clear
                geometricShapesView.borderWidth = model.geometricShapesBorderWidth
                geometricShapesView.backColor = model.geometricShapesBackColor ?? .white
            }
        }
    }
    
    @IBOutlet weak var dropButtonWidthConstraint: NSLayoutConstraint!
    
    @IBOutlet private weak var selectionBtn: UIButton?
    @IBOutlet private weak var contentContainer: UIView!
    
    @IBOutlet private weak var colorBtn: SKUBrightnessBorderButton!
    @IBOutlet private weak var boldButton: UIButton!
    @IBOutlet private weak var dropButton: SKUDropdownButton!
    @IBOutlet private weak var toggleSwitch: UISwitch!
    @IBOutlet private weak var textInputButton: UIButton!
    
    @IBOutlet private weak var geometricShapesView: GeometricShapesView!
    
    // MARK: - Override
    
    override func awakeFromNib() {
        super.awakeFromNib()
        if let selectionBtn = self.selectionBtn {
            selectionBtn.titleLabel?.adjustsFontSizeToFitWidth = true
            selectionBtn.titleLabel?.minimumScaleFactor = 0.5
            
            selectionBtn.tintAdjustmentMode = .normal
            selectionBtn.isUserInteractionEnabled = false
            selectionBtn.translatesAutoresizingMaskIntoConstraints = false
            let constraint = selectionBtn.heightAnchor.constraint(equalToConstant: 44.0)
            constraint.priority = .defaultHigh
            constraint.isActive = true
            selectionBtn.setTitleColor(.black, for: .normal)
        }
        commonInit()
    }
    
    // MARK: - 생성 & 소멸
    
    private func commonInit() {
        contentContainer?.backgroundColor = .clear
        if reuseIdentifier == LineSettingCellID.colorFill.rawValue {
            setupColorFillMode()
        } else if reuseIdentifier == LineSettingCellID.color.rawValue {
            setupColorMode()
        } else if reuseIdentifier == LineSettingCellID.colorDrop.rawValue {
            setupColorDropMode()
        } else if reuseIdentifier == LineSettingCellID.toggle.rawValue {
            setupToggleMode()
        } else if reuseIdentifier == LineSettingCellID.textInput.rawValue {
            setupTextInputMode()
        } else if reuseIdentifier == LineSettingCellID.drop.rawValue {
            setupDropMode()
        } else if reuseIdentifier == LineSettingCellID.font.rawValue {
            setupFontMode()
        } else if reuseIdentifier == LineSettingCellID.geometricShapes.rawValue {
            setupGeometricShapesMode()
        }
    }
    
    private func setupColorFillMode() {
        colorBtn.backgroundColor = .systemMint
        colorBtn.layer.cornerRadius = 4.0
        selectionBtn?.isUserInteractionEnabled = true
        
        guard let selectionBtn = selectionBtn
        else {
            return
        }
        
        if #available(iOS 13, *) {
            if #available(iOS 15, *) {
                var configuration = selectionBtn.configuration
                configuration?.baseBackgroundColor = .clear
                selectionBtn.configuration = configuration
            }
            //
            // let config = UIImage.SymbolConfiguration(scale: .medium)
            // selectedImage = UIImage(systemName: "checkmark.circle.fill", withConfiguration: config)
            // normalImage = UIImage(systemName: "checkmark.circle", withConfiguration: config)
        } else {
            selectionBtn.imageView?.contentMode = .center
            let tintColor = UIColor(red: 15.0/255.0, green: 40.0/255.0, blue: 71.0/255.0, alpha: 1.0)
            selectionBtn.tintColor = tintColor
            selectionBtn.imageView?.tintColor = tintColor
            //
            // let designTimeBundle = Bundle(for: type(of: self)) // Bundle(for: classForCoder) 이렇게도 가능할 듯
            // selectedImage = UIImage(named: "checkmark.circle.fill", in: designTimeBundle, compatibleWith: nil)
            // normalImage = UIImage(named: "checkmark.circle", in: designTimeBundle, compatibleWith: nil)
            // selectedImage = selectedImage?.withRenderingMode(.alwaysTemplate)
            // normalImage = normalImage?.withRenderingMode(.alwaysTemplate)
        }
        let designTimeBundle = Bundle(for: Self.self)
        var selectedImage = UIImage(named: "checkmark.circle.fill", in: designTimeBundle, compatibleWith: nil)
        selectedImage = selectedImage?.withRenderingMode(.alwaysTemplate)
        var normalImage = UIImage(named: "uncheckmark.circle", in: designTimeBundle, compatibleWith: nil)
        normalImage = normalImage?.withRenderingMode(.alwaysTemplate)
        selectionBtn.setImage(selectedImage, for: .selected)
        selectionBtn.setImage(normalImage, for: .normal)
        selectionBtn.setTitleColor(.black, for: .selected)
        selectionBtn.setTitleColor(.lightGray, for: .normal)
    }
    
    private func setupColorMode() {
        colorBtn.backgroundColor = .systemMint
        colorBtn.layer.cornerRadius = 4.0
    }

    private func setupColorDropMode() {
        colorBtn.backgroundColor = .systemMint
        colorBtn.layer.cornerRadius = 4.0
        dropButton.backgroundColor = .clear
        
        dropButton.cellClass = SKULineWidthDropdownCell.self
        dropButton.dropdownData = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
        dropButton.selectedIndex = 0
    }

    private func setupToggleMode() {
    }

    private func setupTextInputMode() {
        textInputButton.backgroundColor = .white
        textInputButton.layer.borderColor = UIColor.lightGray.cgColor
        textInputButton.layer.borderWidth = 1.0 / UIScreen.main.scale
        textInputButton.layer.cornerRadius = 5.0
        textInputButton.setTitleColor(.systemBlue, for: .normal)
        textInputButton.setTitleColor(.gray, for: .disabled)
        textInputButton.titleLabel?.adjustsFontSizeToFitWidth = true
        textInputButton.titleLabel?.minimumScaleFactor = 0.5
    }

    private func setupDropMode() {
        dropButton.backgroundColor = UIColor.clear
        dropButton.cellClass = SKUTextDropdownCell.self
        dropButton.dropdownData = ["목업0", "목업1", "목업2"]
        dropButton.selectedIndex = 1
        dropButton.placeHolderTextAlignment = .center
        dropButton.textAlignment = .right
    }

    private func setupFontMode() {
        boldButton.backgroundColor = .clear
        boldButton.layer.cornerRadius = 4.0
        dropButton.backgroundColor = .clear
        dropButton.cellClass = SKUTextDropdownCell.self
        
        var dropdownData = [String]()
        for i in 10...50 {
            dropdownData.append("\(i)")
        }
        dropButton.dropdownData = dropdownData
        dropButton.layer.cornerRadius = 4.0
        dropButton.selectedIndex = 7
        dropButton.placeHolderTextAlignment = .center
        dropButton.textAlignment = .center
    }

    private func setupGeometricShapesMode() {
    }
}
